torch/csrc/jit/dynamic_dag.h

```diff
@@ -0,0 +1,388 @@
+  auto out_edges = std::move(v->out_edges);
+  vertices_[v->ord] = at::nullopt;
+  delete v;
+  return std::make_pair<vertex_set<T>,vertex_set<T>>(std::move(in_edges), std::move(out_edges));
+}
+
+// This is O(|AR|) which is the same thing as O(ord(consumer) - ord(producer))
+// AR is the "affected region": { v s.t. ord(v) in [ord(producer), ord(consumer)] }
+// consisting of the only verticies that can possibly be moved around due
+// to this edge addition.
+template <typename T>
+void DynamicDAG<T>::addEdge(Vertex<T>* producer, Vertex<T>* consumer) {
+  JIT_ASSERT(producer != consumer);
+  if (producer->out_edges.count(consumer)) return;
+  producer->out_edges.insert(consumer);
+  consumer->in_edges.insert(producer);
+
+  if (producer->ord <= consumer->ord) {
+    // topological ordering is already consistent, no need to update.
+    return;
+  }
+
+  // Search for vertices that are reachable from consumer that have a now incorrect
+  // topological ordering.
+  if (!dfsForward(consumer, producer->ord, deltaF_)) {
+    // dfsForward returns false if there is a path from consumer to producer.
+    throw std::runtime_error("Cycle detected while trying to add edge.");
+  }
+
+  // Search for vertices that can reach producer that have a now incorrect
+  // topological ordering
+  dfsBackward(producer, consumer->ord, deltaB_);
+
+  // Reorder the vertices that are reachable from consumer to occur BEFORE
+  // the vertices that can reach producer.
+  reorder(deltaF_, deltaB_);
+}
+
+// Define the affected region for contractEdge(producer, consumer) as
+// { v s.t. ord(v) in [ord(producer), ord(consumer)] }.
+// These are the only vertices that can possibly be moved around
+// during edge contraction.
+//
+// contractEdge is O(|AR| * |out_edges(producer)|)
+template <typename T>
+bool DynamicDAG<T>::contractEdge(Vertex<T>* producer, Vertex<T>* consumer) {
+  JIT_ASSERT(producer != consumer);
+  if (contractionProducesCycle(producer, consumer)) {
+    return false;
+  }
+
+  removeEdge(producer, consumer);
+
+  // We can either merge producer into consumer or the other way around.
+  // I've chosen to merge producer into consumer for now but this is arbitrary.
+  consumer->rdata.insert(consumer->rdata.end(), producer->rdata.begin(), producer->rdata.end());
+
+  vertex_set<T> in_edges;
+  vertex_set<T> out_edges;
+  std::tie(in_edges, out_edges) = removeVertex(producer);
+
+  // Each of these are constant b/c ord(consumer) > ord(producer) > ord(parent)
+  // so the edge addition still preserves the existing topological order.
+  for (auto* parent : in_edges) {
+    addEdge(parent, consumer);
+  }
+
+  // NB: each addEdge call is linear in (ord(consumer) - ord(child)).
+  // This makes contractEdges O(|out_edges(producer)| * |AR|).
+  for (auto* child : out_edges) {
+    addEdge(consumer, child);
+  }
+  return true;
+}
+
+template <typename T>
+bool DynamicDAG<T>::contractionProducesCycle(Vertex<T>* producer, Vertex<T>* consumer) {
```

Comment:
I feel like this is slightly unnecessary. Couldn't we simply do the `forwardDFS`/`backwardDFS`, and throw if we reach the other end of the edge?

Reply:
Let's say we have x -> y and we're doing a dfs forward from x. The problem is that in its current state dfs forward is going to try to traverse the x -> y edge and then return immediately (or throw). I could have the dfs forward ignore the x -> y edge, but then it wouldn't really be called a dfs, or I can rewrite contractionProducesCycle in terms of dfs

Reply:
I rewrote contractionProducesCycle in terms of DFS for better code reuse